src/tools/storage-key-recipe-resolver.ts

```diff
@@ -81,10 +80,11 @@ export class StorageKeyRecipeResolver {
    * @param opts contains `errors` map for reporting.
    */
   async tryResolve(recipe: Recipe, arc: Arc, opts?: IsValidOptions): Promise<Recipe | null> {
-    const normalized = recipe.clone();
-    const successful = normalized.normalize(opts);
-    if (!successful) return null;
-    if (normalized.isResolved()) return normalized;
+    if (!recipe.normalize(opts)) {
+      throw new StorageKeyRecipeResolverError(
+          `Failed to normalize recipe ${recipe.toString()} with errors:\n${[...opts.errors.values()].join('\n')}`);
+    }
+    if (recipe.isResolved()) return recipe;
 
     return await (new RecipeResolver(arc).resolve(recipe, opts));
```

Comment:
If errors are passed into this `opts` object, we will drop them.

Reply:
moved the null checking and exception throwing inside this method. i think it is better in general.